src/Chapter3.hs

```diff
@@ -1006,6 +1170,143 @@ Implement data types and typeclasses, describing such a battle between two
+
+  doAction :: (Fighter b) => Monster -> b -> Action -> (Monster, b)
+  doAction monster opponent RunAway =
+    ( monster
+    , opponent
+    )
+  doAction monster opponent Attack =
+    ( monster
+    , takeDamage opponent $ monsterAttack monster
+    )
+
+  doActions :: (Fighter a) => Monster -> a -> (Monster, a)
+  doActions monster opponent = head $ map (doAction monster opponent) $ monsterActions monster
+
+instance Fighter Knight where
+  takeDamage :: Knight -> Int -> Knight
+  takeDamage knight amount = knight
+    { knightHealth = knightHealth knight - amount + knightDefense knight
+    , knightDefense = maximum [knightDefense knight - amount, 0]
+    }
+
+  heal :: Knight -> Int -> Knight
+  heal knight amount = knight
+    { knightHealth = knightHealth knight + amount
+    }
+
+  applySpell :: Knight -> Int -> Knight
+  applySpell knight amount = knight
+    { knightDefense = knightDefense knight + amount
+    }
+
+  isDead :: Knight -> Bool
+  isDead = (<= 0) . knightHealth
+
+  attack :: (Fighter a) => Knight -> a -> a
+  attack knight opponent = takeDamage opponent $ knightAttack knight
+
+  doAction :: (Fighter a) => Knight -> a -> Action -> (Knight, a)
+  doAction knight opponent Attack =
+    ( knight
+    , takeDamage opponent $ knightAttack knight
+    )
+  doAction knight opponent DrinkPotion =
+    ( heal knight $ knightPotion knight
+    , opponent
+    )
+  doAction knight opponent CastSpell =
+    ( applySpell knight $ knightSpell knight
+    , takeDamage opponent $ knightAttack knight
+    )
+
+  doActions :: (Fighter a) => Knight -> a -> (Knight, a)
+  doActions knight opponent = head $ map (doAction knight opponent) $ knightActions knight
+
+newtype KnightHealth = KnightHealth { knightHealthValue :: Int }
+newtype KnightAttack = KnightAttack { knightAttackValue :: Int }
+newtype KnightDefense = KnightDefense { knightDefenseValue :: Int }
+newtype Potion = Potion { potionStrength :: Int } deriving (Show)
+newtype Spell = Spell { spellStrength :: Int } deriving (Show)
+
+makeKnight :: KnightHealth -> KnightAttack -> KnightDefense -> Potion -> Spell -> Knight
+makeKnight health attack defense potion spell = Knight
+  { knightHealth   = knightHealthValue health
+  , knightAttack   = knightAttackValue attack
+  , knightDefense  = knightDefenseValue defense
+  , knightActions  = [Attack, CastSpell, DrinkPotion]
+  , knightPotion   = potionStrength potion
+  , knightSpell    = spellStrength spell
+  }
+
+newtype MonsterHealth = MonsterHealth { monsterHealthValue :: Int }
+newtype MonsterAttack = MonsterAttack { monsterAttackValue :: Int }
+
+makeMonster :: MonsterHealth -> MonsterAttack -> Monster
+makeMonster health attack = Monster
+  { monsterHealth  = monsterHealthValue health
+  , monsterAttack  = monsterAttackValue attack
+  , monsterActions = [Attack, RunAway]
+  }
+
+-- Return Int for which player won
+doFight :: (Fighter p1, Fighter p2) => p1 -> p2 -> Int
```

Comment:
`Int` is a too wide type to represent this information. I think it's better to introduce a custom sum type:

```haskell
data BattleResult
    = First
    | Second
```